### ISSUE TITLE
How can we make hub more embeddable

### ISSUE THREAD

**Issue by michael** [substance/hub#1]
The current setup suggests that hub is a stand-alone application. However we want it more to be a library to be pulled into projects that need a doc storage facility. I tried this now with the notepad application so that i just have substance-hub as dependency and start it along the server.

https://github.com/substance/notepad/blob/master/server.js#L20

What feels weird is that i now have to `cd node_modules/substance-hub` and do `npm run seed`. which creates a sqlite file within node_modules/substance-hub/mydb.sqlite. I would want to have this in my project root folder instead. Also what if I want to provide my own seed for the notepad app? where to put it?

Somehow it's really nice that you just `require('substance-hub')` and you get a storage engine and don't need to mess around with database frameworks. For Substance docs we should probably invent our own seeding strategy (list of changes) in our own format (not using knex migrations in substance app development). Then we can also switch out knex with something else.. or use a NOSQL database.

From a usage point of view I want to do this:

in server.js

``` js
var hub = require('substance-hub');

hub.init(app, {
  port: 5000,
  dbFile: './hub-dev.sqlite' // db file gets created automatically if not already exists
});
```

Then hub exposes the low level API which the app developer can use for seeding etc. hub.createChangeset() hub.addChanges usw.

**Comment by Integral** [comment]
Hey!
First of all, you don't need to change directory and execute script. You can do it with npm script.
However I 100% agree with you that we should keep db on application level. And probably we should send more information to hub regarding connection, so we could use not only sqlite, but also other knex-compatible databases. I could do that, should I..?
However there is other significant topic related to hubs... We thought about hub as kind of network unit, but what we are trying to chief here is predefined server side storage. And I don't like it, because it's not allow us to add any additional stuff (like authentication). Any serious app will need more...
Maybe we should consider this thing, what we have here, as pluggable data-receiver for realtime? And it must be extensible somehow...

**Comment by michael** [comment]
I think we need to set priorities.  It's hard to model a full-fledged system with one hub implementation. As you said, people want to use their own stuff.. (not only substance server plugins but let the system integrate with other infrastructure (Elastic Search etc.).

I think it's not a bad idea to define a simple Substance storage layer and dispatcher for realtime collaboration. There must be hooks to implement custom strategies... for instance you could then guard the dispatching mechanism and check if the user has permission to actually modify that document. But yeah that is often very custom stuff.. and given how little time we have we can not implement all this for the hub right now. So let's do the basics on a storage level for now.. and try to anticipate some useful customization hooks.

I'm sorry the whole thing is still not at all clear to me so I can't give you advice to implement certain things. I will try to setup a very basic realtime app using the old operations and without transforming conflicting changes. We will need to replace all this with Oli's latest code.

The first shot will not be a beauty.. but once we have that we can look and see the whole thing and then we probably know what needs to be done to make it better.

**Comment by Integral** [comment]
Yes, we need a hook system... I will think more, at least we also should keep in mind entities case, just to not block that thing.
